R/data.table.R

```diff
@@ -2291,25 +2291,10 @@ transform.data.table = function (`_data`, ...)
 # basically transform.data.frame with data.table instead of data.frame, and retains key
 {
   if (!cedta()) return(NextMethod()) # nocov
-  e = eval(substitute(list(...)), `_data`, parent.frame())
-  tags = names(e)
-  inx = chmatch(tags, names(`_data`))
-  matched = !is.na(inx)
-  if (any(matched)) {
-    .Call(C_unlock, `_data`) # fix for #1641, now covered by test 104.2
-    `_data`[,inx[matched]] = e[matched]
-    `_data` = as.data.table(`_data`)
-  }
-  if (!all(matched)) {
-    ans = do.call("data.table", c(list(`_data`), e[!matched]))
-  } else {
-    ans = `_data`
-  }
-  key.cols = key(`_data`)
```

Comment:
note this behavior for keys, can we double-check it's still working as expected? is there an existing test?

Reply:
There are existing tests. Also note, even if there were a test and we started failing, I would prefer using the new method. The original method was interesting to say the least. The new method should behave as expected. That is, I do not believe you would think twice about key behavior with `dt[, update_stuff := new_stuff]` because we have tests for it. That's what this PR is doing. 

Existing tests:

``` r 
# transform maintains keys
dt <- data.table(a=sample(c('a', 'b', 'c'), 20, replace=TRUE),
                 b=sample(c('a', 'b', 'c'), 20, replace=TRUE),
                 c=sample(20), key=c('a', 'b'))
t1 <- transform(dt, d=c+4)
test(1913.21, key(t1), key(dt))
test(1913.22, t1$d, dt$c + 4)  # transform was successful
t2 <- transform(dt, d=c+4, a=sample(c('x', 'y', 'z'), 20, replace=TRUE))
test(1913.23, is.null(key(t2)))  # transforming a key column nukes the key
## This is probably not necessary, but let's just check that transforming
## a key column doesn't twist around the rows in the result.
test_no = 1913.24
for (col in c('b', 'c')) {
  test_no = test_no + 0.0001
  test(test_no, t2[[col]], dt[[col]])  # mutating-key-transform maintains other columns
}
```